cdcv2/metadata/sql/observation.go

```diff
@@ -0,0 +1,670 @@
+		ownerChanges:     chann.NewAutoDrainChann[metadata.ScheduledChangefeed](),
+		processorChanges: chann.NewAutoDrainChann[metadata.ScheduledChangefeed](),
+	}, nil
+}
+
+// Run runs the given CaptureOb.
+func (c *CaptureOb[T]) Run(
+	egCtx context.Context,
+	controllerCallback func(context.Context, metadata.ControllerObservation) error,
+) (err error) {
+	eg, egCtx := errgroup.WithContext(egCtx)
+	c.egCtx = egCtx
+
+	eg.Go(func() error {
+		return c.Elector.RunElection(egCtx, c.onTakeControl(controllerCallback))
+	})
+
+	eg.Go(func() error {
+		ticker := time.NewTicker(500 * time.Millisecond)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-egCtx.Done():
+				err := egCtx.Err()
+				log.Info("capture stops handle task changes", zap.String("capture", c.selfInfo.ID), zap.Error(err))
+				return err
+			case <-ticker.C:
+				if err := c.handleTaskChanges(egCtx); err != nil {
+					log.Warn("capture handle task changes fail", zap.String("capture", c.selfInfo.ID), zap.Error(err))
+					return err
+				}
+			}
+		}
+	})
+
+	// TODO: add runWithEg function to reduce the wait goroutine.
+	return eg.Wait()
+}
+
+func (c *CaptureOb[T]) onTakeControl(
+	controllerCallback func(context.Context, metadata.ControllerObservation) error,
+) func(context.Context) error {
+	return func(ctx context.Context) error {
+		checker, ok := c.Elector.(LeaderChecker[T])
+		if !ok {
+			return errors.New("capture elector is not a leader checker")
+		}
+		controllerOb := newControllerObservation(checker, c.client, c.selfInfo, c.getAllCaptures)
+
+		eg, egCtx := errgroup.WithContext(ctx)
+		eg.Go(func() error {
+			return controllerOb.run(egCtx)
+		})
+		eg.Go(func() error {
+			return controllerCallback(egCtx, controllerOb)
+		})
+		return eg.Wait()
+	}
+}
+
+func (c *CaptureOb[T]) handleTaskChanges(ctx context.Context) error {
+	var err error
+	var schedItems []*ScheduleDO
+
+	err = c.client.Txn(ctx, func(tx T) error {
+		lastSafePoint := c.tasks.getSafePoint()
+		schedItems, err = c.client.querySchedulesByOwnerIDAndUpdateAt(tx, c.Self().ID, lastSafePoint)
+		return err
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if len(schedItems) == 0 {
+		// No scheudle information for the capture.
+		return nil
+	}
+
+	c.tasks.doUpsert(schedItems, func(newV *ScheduleDO) (skip bool) {
+		if newV.OwnerState == metadata.SchedRemoved {
+			return true
+		}
+		c.ownerChanges.In() <- newV.ScheduledChangefeed
+		c.processorChanges.In() <- newV.ScheduledChangefeed
+		return false
+	})
+
+	return nil
+}
+
+func (c *CaptureOb[T]) Advance(cp metadata.CaptureProgress) error {
+	return c.client.Txn(c.egCtx, func(tx T) error {
+		return c.client.updateProgress(tx, &ProgressDO{
+			CaptureID: c.selfInfo.ID,
+			Progress:  &cp,
+		})
+	})
+}
+
+func (c *CaptureOb[T]) OwnerChanges() <-chan metadata.ScheduledChangefeed {
+	return c.ownerChanges.Out()
+}
+
+func (c *CaptureOb[T]) PostOwnerRemoved(cf metadata.ChangefeedUUID, taskPosition metadata.ChangefeedProgress) error {
```

Comment:
update the metadata interface `PostOwnerRemoved(cf metadata.ChangefeedUUID, taskPosition metadata.ChangefeedProgress) error`